File: Labs/Lab 1/webUI-test/src/test/java/com/accessHQ/models/PlanetPage.java
package com.accessHQ.models;

import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Predicate;

public class PlanetPage {
    private final ChromeDriver driver;

    public PlanetPage(ChromeDriver driver) {
        this.driver = driver;
    }

    public void clickExplore(Planet planet) {
        planet.clickExploreButton();
    }

    public void getPlanet(Predicate<Planet> strategy){
        for(Planet planet: getPlanets()){
            if(strategy.test(planet)){
                clickExplore(planet);
                waitForPopupMessage();
                break;
            }
        }
    }
    private List<Planet> getPlanets() {
        var planets = new ArrayList<Planet>();
        for (WebElement planetWebElement: driver.findElements(By.className("planet"))) {
            planets.add(new Planet(planetWebElement));
        }
        return planets;
    }

    public void waitForPopupMessage() {
        new WebDriverWait(driver,1).until(d->
                ExpectedConditions.visibilityOf(d.findElement(By.className("popup-message"))));
    }

    public String getPopUp() {
        return driver.findElement(By.className("popup-message")).getText();
    }

}

File: Labs/Lab 2/AccessHQ.NETTestProject/Models/PlanetPage.cs
using OpenQA.Selenium;
using OpenQA.Selenium.Support.UI;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AccessHQ.NETTestProject.Models
{
    internal class PlanetPage
    {
        private WebDriver driver;

        public PlanetPage(WebDriver driver)
        {
            this.driver = driver;
        }

        public void clickExplore(Predicate<Planet> strategy)
        {
            foreach (Planet planet in GetPlanets())
            {
                if (strategy.Invoke(planet))
                {
                    clickExploreButton(planet);
                }
            }
        }

        private void clickExploreButton(Planet planet)
        {
            planet.ClickExplore();
        }

        public WebElement getPopUpElement()
        {
            return (WebElement)driver.FindElement(By.ClassName("popup-message"));
        }
        public void waitForPopUp()
        {
            var popup = driver.FindElement(By.ClassName("popup-message"));
            new WebDriverWait(driver, TimeSpan.FromSeconds(10)).Until(d=>
                d.FindElement(By.ClassName("popup-message")).Displayed);
        }

        public List<Planet> GetPlanets()
        {
            List<Planet> planets = new List<Planet>();
            
             foreach(WebElement currentWebElement in driver.FindElements(By.ClassName("planet")))
            {
                planets.Add(new Planet(currentWebElement)); 
            }

            return planets;

        }

    }
}
